src/matcher.py

```diff
@@ -13,58 +16,118 @@
-    environment = os.environ["ENVIRONMENT"]
-    dirty_bucket_name = f"tdr-upload-files-cloudfront-dirty-{environment}"
     s3_client = boto3.client("s3")
-    s3_resource = boto3.resource("s3")
 
+    settings = build_settings(event)
+    download_file_if_not_already_present(settings)
     rules = yara.load("output")
-    efs_root_location = os.environ["ROOT_DIRECTORY"]
+    matched_antivirus_rules = [x.rule for x in rules.match(settings.local_download_location)]
 
-    root_path = f"{efs_root_location}/{consignment_id}"
-    file_path = f"{root_path}/{original_path}"
-    download_directory = "/".join(file_path.split("/")[:-1])
-    if not exists(download_directory):
-        os.makedirs(download_directory)
-    if not exists(file_path):
-        bucket = s3_resource.Bucket(dirty_bucket_name)
-        bucket.download_file(f"{user_id}/{consignment_id}/{file_id}", file_path)
+    if len(matched_antivirus_rules) > 0:
+        if settings.s3_quarantine_location is not None:
+            s3_client.copy(
+                settings.s3_source_location.as_dict(),
+                settings.s3_quarantine_location.bucket,
+                settings.s3_quarantine_location.key
+            )
+    else:
+        if settings.s3_upload_location is not None:
+            s3_client.copy(
+                settings.s3_source_location.as_dict(),
+                settings.s3_upload_location.bucket,
+                settings.s3_upload_location.key
+            )
+
+    logger.info("Key %s processed", settings.s3_source_location.key)
+
+    return antivirus_results_dict(settings.file_id, matched_antivirus_rules, handler_trigger_time)
 
-    match = rules.match(f"{root_path}/{original_path}")
-    results = [x.rule for x in match]
 
-    original_s3_key = f"{user_id}/{consignment_id}/{file_id}"
-    copy_s3_key = f"{consignment_id}/{file_id}"
+@dataclass(frozen=True)
+class S3Location:
+    bucket: str
+    key: str
 
-    copy_source = {
-        "Bucket": dirty_bucket_name,
-        "Key": original_s3_key
+    def as_dict(self) -> object: return {
+        "Bucket": self.bucket,
+        "Key": self.key
     }
 
-    if len(results) > 0:
-        s3_client.copy(
-            copy_source,
-            "tdr-upload-files-quarantine-" + environment,
-            copy_s3_key
+
+@dataclass(frozen=True)
+class VirusCheckSettings:
+    file_id: str
+    s3_source_location: S3Location
+    s3_quarantine_location: Optional[S3Location]
+    s3_upload_location: Optional[S3Location]
+    local_download_location: str
+
+
+class ScanType(Enum):
+    metadata = 1
+    consignment = 2
+
+
+def build_settings(event: dict) -> VirusCheckSettings:
+    scan_type = ScanType[event.get("scanType", "consignment")]
+    consignment_id = event["consignmentId"]
+    file_id = event["fileId"]
+
+    environment = os.environ["ENVIRONMENT"]
+    efs_root_location = os.environ["ROOT_DIRECTORY"]
+    root_path = f"{efs_root_location}/{consignment_id}"
+    if scan_type == ScanType.consignment:
```

Comment:
I would switch this around and have the metadata scan as the first, and then an else statement to cover everything else. eg:

```
if scan_type == ScanType.metadata:
...
else:
...
```